src/coreclr/src/jit/lowerxarch.cpp

```diff
@@ -2451,6 +2520,702 @@ void Lowering::LowerHWIntrinsicCreate(GenTreeHWIntrinsic* node)
+                                                          simdSize);
+                    BlockRange().InsertAfter(idx, tmp3);
+                    LowerNode(tmp3);
+
+                    node->gtOp1 = tmp3;
+                    node->gtOp2 = nullptr;
+
+                    node->gtHWIntrinsicId = NI_Vector128_ToScalar;
+                    LowerNode(node);
+
+                    return;
+                }
+
+                multiply      = NI_SSE_Multiply;
+                horizontalAdd = NI_SSE3_HorizontalAdd;
+                add           = NI_SSE_Add;
+
+                if (!comp->compOpportunisticallyDependsOn(InstructionSet_SSE3))
+                {
+                    shuffle = NI_SSE_Shuffle;
+                }
+                break;
+            }
+
+            case TYP_DOUBLE:
+            {
+                if (comp->compOpportunisticallyDependsOn(InstructionSet_SSE41))
+                {
+                    // We will be constructing the following parts:
+                    //   idx  =    CNS_INT       int    0x31
+                    //          /--*  op1  simd16
+                    //          +--*  op2  simd16
+                    //          +--*  idx  int
+                    //   tmp3 = *  HWINTRINSIC   simd16 T DotProduct
+                    //          /--*  tmp3 simd16
+                    //   node = *  HWINTRINSIC   simd16 T ToScalar
+
+                    // This is roughly the following managed code:
+                    //   var tmp3 = Avx.DotProduct(op1, op2, 0x31);
+                    //   return tmp3.ToScalar();
+
+                    idx = comp->gtNewIconNode(0x31, TYP_INT);
+                    BlockRange().InsertBefore(node, idx);
+
+                    tmp3 = comp->gtNewSimdHWIntrinsicNode(simdType, op1, op2, idx, NI_SSE41_DotProduct, baseType,
+                                                          simdSize);
+                    BlockRange().InsertAfter(idx, tmp3);
+                    LowerNode(tmp3);
+
+                    node->gtOp1 = tmp3;
+                    node->gtOp2 = nullptr;
+
+                    node->gtHWIntrinsicId = NI_Vector128_ToScalar;
+                    LowerNode(node);
+
+                    return;
+                }
+
+                multiply      = NI_SSE2_Multiply;
+                horizontalAdd = NI_SSE3_HorizontalAdd;
+                add           = NI_SSE2_Add;
+
+                if (!comp->compOpportunisticallyDependsOn(InstructionSet_SSE3))
+                {
+                    shuffle = NI_SSE2_Shuffle;
+                }
+                break;
+            }
+
+            default:
+            {
+                unreached();
+            }
+        }
+
+        if (simdSize == 8)
+        {
+            assert(baseType == TYP_FLOAT);
+
+            // Setting the simd16Count to 2 will cause 1 hadd to be emitted rather
+            // than 2, which will cause the upper two elements to be ignored.
```

Comment:
This is somewhat confusing. I would probably say
```suggestion
            // If simdSize == 8 then we have only two elements, not the 4 that we got from getSIMDVectorLength,
            // which we gave a simdSize of 16. So, we set the simd16Count to 2 so that only 1 hadd will
            // be emitted rather than 2, so that the upper two elements will be ignored.
```
Of course, please correct if I've got this wrong!